command/metadata.go

```diff
@@ -0,0 +1,142 @@
+package command
+
+import (
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/mackerelio/mackerel-agent/config"
+	"github.com/mackerelio/mackerel-agent/metadata"
+)
+
+func metadataGenerators(conf *config.Config) []*metadata.Generator {
+	generators := make([]*metadata.Generator, 0, len(conf.MetadataPlugins))
+
+	workdir := os.Getenv("MACKEREL_PLUGIN_WORKDIR")
+	if workdir == "" {
+		workdir = os.TempDir()
+	}
+
+	for name, pluginConfig := range conf.MetadataPlugins {
+		generator := &metadata.Generator{
+			Name:      name,
+			Config:    pluginConfig,
+			Cachefile: filepath.Join(workdir, "mackerel-metadata", name),
+		}
+		logger.Debugf("Metadata plugin generator created: %#v %#v", generator, generator.Config)
+		generators = append(generators, generator)
+	}
+
+	return generators
+}
+
+type metadataResult struct {
+	namespace string
+	metadata  interface{}
+	createdAt time.Time
+}
+
+func runMetadataLoop(c *Context, termMetadataCh <-chan struct{}, quit <-chan struct{}) {
+	resultCh := make(chan *metadataResult)
+	for _, g := range c.Agent.MetadataGenerators {
+		go runEachMetadataLoop(g, resultCh, quit)
+	}
+
+	exit := false
+	for !exit {
+		select {
+		case <-time.After(1 * time.Minute):
+		case <-termMetadataCh:
+			logger.Debugf("received 'term' chan for metadata loop")
+			exit = true
+		}
+
+		results := make(map[string]*metadataResult)
+	ConsumeResults:
+		for {
+			select {
+			case result := <-resultCh:
+				// prefer new result to avoid infinite number of retries
+				if prev, ok := results[result.namespace]; ok {
```

Comment:
Thanks, I'll keep the code because I feel the if statement a kind of an idiom in Golang.